### ISSUE TITLE
Memory leak(?) while persisting models in memory after training

### ISSUE THREAD

**Issue by yupopov** [catboost/catboost#2827]
catboost version: 1.2.7
Operating System: Debian bookworm (python:3.11 docker image)
CPU: AMD EPYC 7502P 32-Core Processor
GPU: N/A

## Problem
I came across a strange behavior: saving a model to memory after training increases memory usage by a seemingly unreasonable amount, and the memory isn't released until the death of current Python process.
Particularly, training and persisting many models in the same runtime eventually leads to an OOM error.
The problem disappears if I persist a copy instead of the model itself.

## Reproduction

- `git clone git@github.com:yupopov/catboost_train_mem_logging.git`
- `docker compose up -d`
- Run the script `log_train_memory_consumption.sh` in the container `cb-mem-logger` with certain parameters, see examples below.

*TL;DR of what it does:* 
- Generate synthetic pools;
- For each pool: train a Catboost model, delete the model / save the model to memory / save model copy in memory, log memory consumption;
- Repeat the previous step a certain number of times because the memory usage may differ from run to run;
- Plot memory consumption increases.

In more detail:
- Run `initialize_experiment.py`: make `m` synthetic pools by `sklearn.datasets.make_classification` (with the same parameters except for different random seeds), randomly add categorical features, save them to disk as `.bin` files.
- Run `log_train_memory_consumption.py` `n` times: 
	- load the pools from disk;
	- for each pool:
		- fit a `CatBoostClassifier` on the pool (no validation set, model and fit params are the same for each model, particularly, `max_ctr_complexity = 1`);
		- delete the pool, run `gc.collect()`;
		- delete the model / save the model to memory / copy the model and then save it to memory (governed by the parameter `-s`);
		- record memory consumption with `psutil`;
	- save memory consumption logs to a file, (optionally) save fitted models to `.cbm` files.
This is the relevant code part:
```
    mem_footprints: list[float] = []
    models: list[tuple[CatBoostClassifier, Path]] = []

    mem_footprints.append(memory_footprint())
    for (model_num, train_pool) in tqdm(enumerate(train_pools), total=len(train_pools)):
        model = CatBoostClassifier(**model_params)
        model.fit(
            X=train_pool,
            **fit_params
        )
        del train_pool
        gc.collect()
	model_path = ...
        if save_models == 'no':
            del model
        if save_models == 'save':
            models.append((model, model_path))
        if save_models == 'copy':
            models.append((model.copy(), model_path))
        gc.collect()
        mem_footprints.append(memory_footprint())
```

- Run `analyze_memory_increases.py`: plot memory increases for each run, save the plot to a `.jpg` file. If models were saved to disk, load all `n * m` models to memory and check by how much the memory usage increases.

Dataset, model and fit parameters can be found in `config.json`, the location of train artifacts is defined by the parameter `-e`.

## Results
Below are the memory increases plots along with the commands to reproduce the result.
In each script run, we train 20 models on the same data 10 times.
Dataset, model in fit parameters are the same everywhere, the only difference is what we do with the models after training.

- Delete the model after each training:
 `docker exec cb-mem-logger bash log_train_memory_consumption.sh -e 200_iter_no_save_models -c config.json -m 20 -n 10 -s no`
![memory increases plot when we delete models after each training](https://raw.githubusercontent.com/yupopov/catboost_train_mem_logging/refs/heads/master/exp_results/200_iter_no_save_models/mem_increases_plot.jpg)

Here the memory consumption increase drops to almost zero with time for all runs.

- Save the models after the training:
 `docker exec cb-mem-logger bash log_train_memory_consumption.sh -e 200_iter_save_models -c config.json -m 20 -n 10 -s save`
![memory increases plot when we save models after each training](https://raw.githubusercontent.com/yupopov/catboost_train_mem_logging/refs/heads/master/exp_results/200_iter_save_models/mem_increases_plot.jpg)
Here memory increases stabilize at around 30-40 Mb with time.
Clearly models should take up a certain amount of memory, but not that much: loading all 10*20 = 200 models in memory after all runs increases memory usage by around 45 Mb.
**This is the behavior I want to avoid.**

- Save model copies after the training:
 `docker exec cb-mem-logger bash log_train_memory_consumption.sh -e 200_iter_save_model_copies -c config.json -m 20 -n 10 -s copy`
![memory increases plot when we save model copies after each training](https://raw.githubusercontent.com/yupopov/catboost_train_mem_logging/refs/heads/master/exp_results/200_iter_save_model_copies/mem_increases_plot.jpg)

This plot is similar to the first one, no constant memory increase.
Problem averted in an indirect way? No idea why this works, would be grateful if someone could explain.

I suppose I'm doing something wrong here, because fitting a bunch of models and persisting them in memory seems a fairly standard workflow, but can't find what exactly is wrong.
Let me know if I can provide more info.

Dataset and model parameters used above are purely exemplary, I observed the same behavior without categorical features, different number of iterations, etc.
For reproducing the results faster you can also decrease the number of runs (parameter `-n`).
